### ISSUE TITLE
Map constructor doesn't work with target ES5, lib ES6

### ISSUE THREAD

**Issue by ivogabe** [Microsoft/TypeScript#8407]
**TypeScript Version:** 

1.7.5 / 1.8.0-beta / nightly (1.9.0-dev.20160430)

**Code**

```ts
const x = new Map([["a", 1]]); // Error
const y = new Map<string, number>([["a", 1]]); // Error
const z = new Map([<[string, number]>["a", 1]]); // Works
```
```json
{
	"compilerOptions": {
		"target": "es5",
		"lib": ["es2016"]
	}
}
```
**Expected behavior:**
No errors

**Actual behavior:** 
Errors are reported on the first two lines:
```
Argument of type '(string | number)[][]' is not assignable to parameter of type 'Iterable<[{}, {}]>'. 
Argument of type '(string | number)[][]' is not assignable to parameter of type 'Iterable<[string, number]>'.
```
I think the reason is that the compiler doesn't use iterators to resolve the type arguments when the target is ES5. Adding an overload to the `Map` constructor, which takes an array of tuples instead of an iterator, would probably fix this. I'm not sure whether this happens on more locations though, if so it might be worth to pick up iterators in the compiler even when the target is ES5.

**Comment by mhegazy** [comment]
Iterator support are not guaranteed to exist if target is not ES6, so i think the assumption we have today is accurate. I agree with your  proposal to add an array overload for `Map`, `WeakMap`, `Set` and `WeakSet`. A PR would be appreciated.

**Comment by ivogabe** [comment]
I've opened PR #8422

**Comment by ivogabe** [comment]
Thanks!
